OpenRA.Mods.Common/Widgets/Logic/GlobalChatLogic.cs

```diff
@@ -0,0 +1,158 @@
+using OpenRA.Chat;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	class GlobalChatLogic : IDisposable
+	{
+		readonly ScrollPanelWidget historyPanel;
+		readonly LabelWidget historyTemplate;
+		readonly ScrollPanelWidget nicknamePanel;
+		readonly Widget nicknameTemplate;
+		readonly TextFieldWidget inputBox;
+
+		[ObjectCreator.UseCtor]
+		public GlobalChatLogic(Widget widget)
+		{
+			historyPanel = widget.Get<ScrollPanelWidget>("HISTORY_PANEL");
+			historyTemplate = historyPanel.Get<LabelWidget>("HISTORY_TEMPLATE");
+			nicknamePanel = widget.Get<ScrollPanelWidget>("NICKNAME_PANEL");
+			nicknameTemplate = nicknamePanel.Get("NICKNAME_TEMPLATE");
+
+			historyPanel.Bind(Game.GlobalChat.History, MakeHistoryWidget, HistoryWidgetEquals, true);
+			nicknamePanel.Bind(Game.GlobalChat.Users, MakeUserWidget, UserWidgetEquals, false);
+
+			inputBox = widget.Get<TextFieldWidget>("CHAT_TEXTFIELD");
+			inputBox.IsDisabled = () => Game.GlobalChat.ConnectionStatus != ChatConnectionStatus.Joined;
+			inputBox.OnEnterKey = EnterPressed;
+
+			var nicknameBox = widget.Get<TextFieldWidget>("NICKNAME_TEXTFIELD");
+			nicknameBox.Text = Game.GlobalChat.SanitizedName(Game.Settings.Chat.Nickname);
+			nicknameBox.OnTextEdited = () =>
+			{
+				nicknameBox.Text = Game.GlobalChat.SanitizedName(nicknameBox.Text);
+			};
+
+			var connectPanel = widget.Get("GLOBALCHAT_CONNECT_PANEL");
+			connectPanel.IsVisible = () => Game.GlobalChat.ConnectionStatus == ChatConnectionStatus.Disconnected;
+
+			var disconnectButton = widget.Get<ButtonWidget>("DISCONNECT_BUTTON");
+			disconnectButton.OnClick = Game.GlobalChat.Disconnect;
+
+			var connectAutomaticallyCheckBox = connectPanel.Get<CheckboxWidget>("CONNECT_AUTOMATICALLY_CHECKBOX");
+			connectAutomaticallyCheckBox.IsChecked = () => Game.Settings.Chat.ConnectAutomatically;
+			connectAutomaticallyCheckBox.OnClick = () => { Game.Settings.Chat.ConnectAutomatically ^= true; Game.Settings.Save(); };
+
+			var connectButton = connectPanel.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.IsDisabled = () => !Game.GlobalChat.IsValidNickname(nicknameBox.Text);
+			connectButton.OnClick = () =>
+			{
+				Game.Settings.Chat.Nickname = nicknameBox.Text;
+				Game.Settings.Save();
+				Game.GlobalChat.Connect();
+			};
+
+			var mainPanel = widget.Get("GLOBALCHAT_MAIN_PANEL");
+			mainPanel.IsVisible = () => Game.GlobalChat.ConnectionStatus != ChatConnectionStatus.Disconnected;
+
+			mainPanel.Get<LabelWidget>("CHANNEL_TOPIC").GetText = () => Game.GlobalChat.Topic;
+
+			if (Game.Settings.Chat.ConnectAutomatically && Game.GlobalChat.IsValidNickname(Game.Settings.Chat.Nickname))
+				Game.GlobalChat.Connect();
+		}
+
+		Widget MakeHistoryWidget(object o)
+		{
+			var message = (ChatMessage)o;
+			var widget = (LabelWidget)historyTemplate.Clone();
+			var font = Game.Renderer.Fonts[widget.Font];
+
+			var color = message.Type == ChatMessageType.Notification ?
+				Color.LightGray : Color.White;
```

Comment:
:+1: I like the subtle color differentiation, but I guess this wants to live in `metrics.yaml` instead.